meta-client/src/test/java/org/apache/helix/metaclient/impl/zk/TestMultiThreadStressTest/TestMultiThreadStressZKClient.java

```diff
@@ -0,0 +1,368 @@
+
+    // cleanup
+    _zkMetaClient.recursiveDelete(zkParentKey);
+    Assert.assertEquals(_zkMetaClient.countDirectChildren(zkParentKey), 0);
+  }
+
+  @Test
+  public void testSetPuppy() {
+    _zkMetaClient.create(zkParentKey, "test");
+
+    PuppySpec puppySpec = new PuppySpec(PuppyMode.REPEAT, 0.2f, new ExecDelay(5000, 0.1f), 5);
+    CreatePuppy createPuppy = new CreatePuppy(_zkMetaClient, puppySpec);
+    SetPuppy setPuppy = new SetPuppy(_zkMetaClient, puppySpec);
+
+    PuppyManager puppyManager = new PuppyManager();
+    puppyManager.addPuppy(createPuppy);
+    puppyManager.addPuppy(setPuppy);
+
+    long timeoutInSeconds = 60; // Set the desired timeout duration
+
+    puppyManager.start(timeoutInSeconds);
+
+    assertNoExceptions(puppyManager, null);
+
+    // cleanup
+    _zkMetaClient.recursiveDelete(zkParentKey);
+    Assert.assertEquals(_zkMetaClient.countDirectChildren(zkParentKey), 0);
+  }
+
+  @Test
+  public void testUpdatePuppy() {
+    _zkMetaClient.create(zkParentKey, "test");
+
+    PuppySpec puppySpec = new PuppySpec(PuppyMode.REPEAT, 0.2f, new ExecDelay(5000, 0.1f), 5);
+    CreatePuppy createPuppy = new CreatePuppy(_zkMetaClient, puppySpec);
+    UpdatePuppy updatePuppy = new UpdatePuppy(_zkMetaClient, puppySpec);
+
+    PuppyManager puppyManager = new PuppyManager();
+    puppyManager.addPuppy(createPuppy);
+    puppyManager.addPuppy(updatePuppy);
+
+    long timeoutInSeconds = 60; // Set the desired timeout duration
+
+    puppyManager.start(timeoutInSeconds);
+
+    assertNoExceptions(puppyManager, null);
+
+    // cleanup
+    _zkMetaClient.recursiveDelete(zkParentKey);
+    Assert.assertEquals(_zkMetaClient.countDirectChildren(zkParentKey), 0);
+  }
+
+  @Test
+  public void testCrudPuppies() {
+    _zkMetaClient.create(zkParentKey, "test");
+
+    PuppySpec puppySpec = new PuppySpec(PuppyMode.REPEAT, 0.2f, new ExecDelay(5000, 0.1f), 5);
+    CreatePuppy createPuppy = new CreatePuppy(_zkMetaClient, puppySpec);
+    GetPuppy getPuppy = new GetPuppy(_zkMetaClient, puppySpec);
+    DeletePuppy deletePuppy = new DeletePuppy(_zkMetaClient, puppySpec);
+    SetPuppy setPuppy = new SetPuppy(_zkMetaClient, puppySpec);
+    UpdatePuppy updatePuppy = new UpdatePuppy(_zkMetaClient, puppySpec);
+
+    PuppyManager puppyManager = new PuppyManager();
+    puppyManager.addPuppy(createPuppy);
+    puppyManager.addPuppy(getPuppy);
+    puppyManager.addPuppy(deletePuppy);
+    puppyManager.addPuppy(setPuppy);
+    puppyManager.addPuppy(updatePuppy);
+
+    long timeoutInSeconds = 60; // Set the desired timeout duration
+
+    puppyManager.start(timeoutInSeconds);
+
+    assertNoExceptions(puppyManager, null);
+
+    // cleanup
+    _zkMetaClient.recursiveDelete(zkParentKey);
+    Assert.assertEquals(_zkMetaClient.countDirectChildren(zkParentKey), 0);
+  }
+
+
+  @Test
+  public void testBasicParentListenerPuppy() {
+    _zkMetaClient.create(zkParentKey, "test");
+    AtomicInteger globalChildChangeCounter = new AtomicInteger();
```

Comment:
Since we are not actually reading and doing incremental at the same time, why we need AtomicInteger?

Reply:
For lambda expressions we need the variable to be final and atomicIntegers works. While it is not necessary to have a lambda expression, the overhead of using an atomicinteger over a regular int is minimal in this scenario and might not be worth the additional lines of code (that make it slightly messier). Let me know if this makes sense, thanks!